File: public/python/063-maximum-product-subarray.py
# LeetCode Problem #152: Maximum Product Subarray
# Difficulty: Medium
# Link: https://leetcode.com/problems/maximum-product-subarray/

from typing import List

# ─────────────────────────────────────────────
# APPROACH 1: Brute Force | O(n²) time | O(1) space
# EXPLAIN: Try every subarray, compute its product, track the maximum seen.
# WHEN: Only for small inputs or when code clarity matters more than performance.

def maxProduct_brute(nums: List[int]) -> int:
    n = len(nums)
    result = nums[0]
    for i in range(n):
        prod = 1
        for j in range(i, n):
            prod *= nums[j]
            result = max(result, prod)
    return result


# APPROACH 2: DP Track Min/Max | O(n) time | O(1) space
# EXPLAIN: At each element keep both the max and min product ending here (negatives can flip);
#          global max is updated at every step.
# WHEN: Optimal — handles negatives and zeros elegantly in a single pass.

def maxProduct_dp(nums: List[int]) -> int:
    cur_max = cur_min = result = nums[0]
    for num in nums[1:]:
        candidates = (num, cur_max * num, cur_min * num)
        cur_max = max(candidates)
        cur_min = min(candidates)
        result = max(result, cur_max)
    return result


# APPROACH 3: Kadane Variant — Swap on Negative | O(n) time | O(1) space
# EXPLAIN: When element is negative swap max/min before updating — cleaner swap formulation.
# WHEN: Alternative O(1) form; slightly more readable when demonstrating the swap insight.

def maxProduct_kadane(nums: List[int]) -> int:
    cur_max = cur_min = result = nums[0]
    for num in nums[1:]:
        if num < 0:
            cur_max, cur_min = cur_min, cur_max
        cur_max = max(num, cur_max * num)
        cur_min = min(num, cur_min * num)
        result  = max(result, cur_max)
    return result


# APPROACH 4: Two-Pass (Left↔Right) | O(n) time | O(1) space
# EXPLAIN: Scan left→right and right→left tracking running product; reset on zero.
# WHEN: Elegant dual-pass that naturally handles even numbers of negatives and zeros.

def maxProduct_two_pass(nums: List[int]) -> int:
    result = max(nums)
    product = 1
    for num in nums:
        product *= num
        result = max(result, product)
        if product == 0:
            product = 1
    product = 1
    for num in reversed(nums):
        product *= num
        result = max(result, product)
        if product == 0:
            product = 1
    return result


# APPROACH 5: Divide by Zeros (segment approach) | O(n) time | O(1) space
# EXPLAIN: Split array at zeros; for each non-zero segment apply DP; zeros contribute max 0.
# WHEN: Makes zero-handling explicit; good for teaching the role each zero plays.

def maxProduct_divide_zeros(nums: List[int]) -> int:
    result = nums[0]
    max_p = min_p = nums[0]
    for num in nums[1:]:
        if num == 0:
            result = max(result, 0)
            max_p = min_p = 0
        else:
            new_max = max(num, max_p * num, min_p * num)
            new_min = min(num, max_p * num, min_p * num)
            max_p, min_p = new_max, new_min
            result = max(result, max_p)
    return result


# ── quick tests ──────────────────────────────────────────────────────────────
if __name__ == '__main__':
    cases = [
        ([2, 3, -2, 4], 6),
        ([-2, 0, -1], 0),
        ([-2], -2),
        ([0, 2], 2),
        ([-2, 3, -4], 24),
        ([2, -5, -2, -4, 3], 24),
    ]
    for fn in (maxProduct_brute, maxProduct_dp, maxProduct_kadane,
               maxProduct_two_pass, maxProduct_divide_zeros):
        for nums, expected in cases:
            result = fn(nums)
            assert result == expected, (
                f'{fn.__name__}({nums}) = {result}, expected {expected}'
            )
    print('All tests passed.')

# Made with Bob

File: public/java/063-maximum-product-subarray.java
/*
 * LeetCode Problem #152: Maximum Product Subarray
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/maximum-product-subarray/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Brute Force | O(n²) time | O(1) space
    // EXPLAIN: Try every subarray start/end pair and track the running product.
    // WHEN: Use only to verify correctness or when n is tiny (≤100).

    public int maxProduct_brute(int[] nums) {
        int result = nums[0];
        for (int i = 0; i < nums.length; i++) {
            int product = 1;
            for (int j = i; j < nums.length; j++) {
                product *= nums[j];
                result = Math.max(result, product);
            }
        }
        return result;
    }

    // APPROACH 2: DP tracking min and max | O(n) time | O(1) space
    // EXPLAIN: At each index keep both the max and min product ending here; a negative flips min↔max.
    // WHEN: Use as the optimal solution — handles negative numbers elegantly in one pass.

    public int maxProduct(int[] nums) {
        int maxProd = nums[0], minProd = nums[0], result = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            int tempMax = Math.max(num, Math.max(maxProd * num, minProd * num));
            int tempMin = Math.min(num, Math.min(maxProd * num, minProd * num));
            maxProd = tempMax;
            minProd = tempMin;
            result = Math.max(result, maxProd);
        }
        return result;
    }

    // APPROACH 3: Kadane Variant (swap on negative) | O(n) time | O(1) space
    // EXPLAIN: When element is negative, swap maxProd and minProd before updating.
    // WHEN: Cleaner swap-based formulation of the same O(1) DP logic.

    public int maxProduct_kadane(int[] nums) {
        int maxProd = nums[0], minProd = nums[0], result = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < 0) { int tmp = maxProd; maxProd = minProd; minProd = tmp; }
            maxProd = Math.max(nums[i], maxProd * nums[i]);
            minProd = Math.min(nums[i], minProd * nums[i]);
            result  = Math.max(result, maxProd);
        }
        return result;
    }

    // APPROACH 4: Two-Pass (Left→Right and Right→Left) | O(n) time | O(1) space
    // EXPLAIN: Scan each direction tracking running product; reset on zero; take overall max.
    // WHEN: Elegant alternative that handles zeros and negatives via dual pass.

    public int maxProduct_twoPass(int[] nums) {
        int n = nums.length, result = nums[0], product = 1;
        // Left to right
        for (int i = 0; i < n; i++) {
            product *= nums[i];
            result   = Math.max(result, product);
            if (product == 0) product = 1;
        }
        // Right to left
        product = 1;
        for (int i = n - 1; i >= 0; i--) {
            product *= nums[i];
            result   = Math.max(result, product);
            if (product == 0) product = 1;
        }
        return result;
    }

    // APPROACH 5: Divide by Zeros | O(n) time | O(1) space
    // EXPLAIN: Split array at zeros; for each non-zero segment apply DP tracking min/max.
    // WHEN: Makes the zero-handling logic explicit; useful for understanding edge cases.

    public int maxProduct_divideZeros(int[] nums) {
        int result = nums[0];
        int maxP = nums[0], minP = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] == 0) {
                result = Math.max(result, 0);
                maxP = 0; minP = 0;
            } else {
                int newMax = Math.max(nums[i], Math.max(maxP * nums[i], minP * nums[i]));
                int newMin = Math.min(nums[i], Math.min(maxP * nums[i], minP * nums[i]));
                maxP = newMax; minP = newMin;
                result = Math.max(result, maxP);
            }
        }
        return result;
    }
}

// Made with Bob
